### PR TITLE
Allow quantumcircuit construction with zero classical registers

### PR DESCRIPTION
**Pull request by peendebak**
### Summary

Allow circuits with zero classical registers to be created in the same way as circuits with classical registers.

### Details and comments

An example
```
from qiskit.circuit import QuantumCircuit
c = QuantumCircuit(2,2)
c = QuantumCircuit(2,0) # was not possible
```

### CODE REVIEW

**Comment by mtreinish** [comment]
So I'm curious is there an advantage to doing this over:
 
```python
from qiskit.circuit import QuantumCircuit
c = QuantumCircuit(2)
```

because that works fine today. Adding support for 0 seems kind of weird to me because we enforce a positive non-zeo int for qubits.

**Comment by Cryoris** [comment]
I could see the value of allowing 0 bits in the initializer, e.g. if you automatically construct and compose circuits and some of them happen to have 0 qubits/clbits. But I agree with @mtreinish, it's a bit strange to only have that for classical bits. If we add this, I think we should have this for both arguments and also for the registers (i.e. `QuantumRegister(0)`).

**Comment by peendebak** [comment]
@mtreinish The case that @Cryoris describes is exactly what we encountered and why this PR was made. 

I updated the PR so that quantum register of size 0 are allows, and I updated the tests.

File: qiskit/circuit/quantumcircuit.py [hunk 1]
```diff
@@ -1151,7 +1151,10 @@ def add_register(self, *regs):
                 regs = (QuantumRegister(regs[0], "q"),)
             elif len(regs) == 2 and all(isinstance(reg, int) for reg in regs):
                 # QuantumCircuit with anonymous wires e.g. QuantumCircuit(2, 3)
-                regs = (QuantumRegister(regs[0], "q"), ClassicalRegister(regs[1], "c"))
+                if regs[1] == 0:
+                    regs = (QuantumRegister(regs[0], "q"),)
+                else:
+                    regs = (QuantumRegister(regs[0], "q"), ClassicalRegister(regs[1], "c"))
```
Comment (ajavadia) [inline qiskit/circuit/quantumcircuit.py:1157]: If size 0 is allowed then this modification is not needed i think

File: qiskit/circuit/register.py [hunk 2]
```diff
@@ -82,7 +82,7 @@ def __init__(self, size=None, name=None, bits=None):
             )
         size = int(size)  # cast to int
 
-        if size <= 0:
+        if size < 0:
             raise CircuitError(
                 "Register size must be positive (%s '%s' was provided)"
```
Comment (ajavadia) [inline qiskit/circuit/register.py:87]: the error should now say "non-negative"

**Review by ajavadia — COMMENTED** [review]
I think it makes sense to allow zero-length registers in general (they're basically like an empty list of Qubits or empty list of Clbits). I certainly prefer this than the stop-gap solution of allowing `QuantumCircuit(n, 0)` to work.

But we need to make sure the rest of Qiskit is prepared to handle circuits circuits with zero-length registers. For example the circuit drawers or the transpiler should not throw errors.

File: qiskit/circuit/register.py (hunk 2 above)
Comment (eendebakpt) [reply qiskit/circuit/register.py]: @ajavadia As a mathematician positive includes zero (and strictly positive does not include zero). But the term non-negative is clearer indeed.

File: qiskit/circuit/quantumcircuit.py (hunk 1 above)
Comment (eendebakpt) [reply qiskit/circuit/quantumcircuit.py:1157]: @ajavadia Removing this `if` statement triggers some issues with the tests. There is a difference between a circuit with no classical registers (i.e. an empty list), or a circuit with a classical register of size 0.

image

The behaviour in this PR. e.g. `QuantumCircuit(2,0)` creates a circuit without classical registers, is consistent with current behaviour (and tests).

**Comment by peendebak** [comment]
@mtreinish Could you have another look at the PR?

**Comment by peendebak** [comment]
@ajavadia Could you have a look at the PR and let me know whether it needs some more improvements?

**Review by kdk — COMMENTED** [review]
Code changes so far LGTM, but I agree with @mtreinish and @Cryoris that this introduces an unusual difference in behavior between `QuantumRegister`s and `ClassicalRegister`s.

After this PR, `QuantumCircuit(0, 0)` would have one 0-length `QuantumRegister`, but no `ClassicalRegister`s. I'd vote for also allowing `QuantumCircuit(0)` and `QuantumCircuit(0, N)` to not create a `QuantumRegister`.

**Comment by eendebakpt** [comment]
@mtreinish @Cryoris @kdk The current behaviour is
image

There is indeed a difference in behaviour between the registers, but it is most consistent with the current behaviour. I don't mind changing to another type of behaviour, but please first come to consensus before I update the PR again.

**Comment by eendebakpt** [comment]
@kdk The behavior you suggested I implemented in #6800

**Comment by kdk** [comment]
Closing in favor of #6800 .

**Pull request closed without merging**